awx/main/analytics/core.py

```diff
@@ -82,145 +61,152 @@ def decorate(f):
+
+        dest = pathlib.Path(dest or tempfile.mkdtemp(prefix='awx_analytics'))
+        gather_dir = dest.joinpath('stage')
+        gather_dir.mkdir(mode=0o700)
+        tarfiles = []
+
+        # These json collectors are pretty compact, so collect all of them before shipping to analytics.
+        data = {}
+        for func in json_collectors:
             key = func.__awx_analytics_key__
-            path = '{}.json'.format(os.path.join(gather_dir, key))
-            with open(path, 'w', encoding='utf-8') as f:
-                try:
-                    json.dump(func(last_run, collection_type=collection_type, until=until), f)
-                    manifest['{}.json'.format(key)] = func.__awx_analytics_version__
-                except Exception:
-                    logger.exception("Could not generate metric {}.json".format(key))
-                    f.close()
-                    os.remove(f.name)
-        elif func.__awx_analytics_type__ == 'csv':
+            filename = f'{key}.json'
+            try:
+                data[filename] = (func(last_run, collection_type=collection_type, until=until),
+                                  func.__awx_analytics_version__)
+            except Exception:
+                logger.exception("Could not generate metric {}".format(filename))
+        if data:
+            tgzfile = package(dest.parent, data, until)
+            if tgzfile is not None:
+                tarfiles.append(tgzfile)
+
+                if collection_type != 'dry-run':
+                    ship(tgzfile)
+                    with disable_activity_stream():
+                        for filename in data:
+                            last_entries[filename.replace('.json', '')] = until
+                        settings.AUTOMATION_ANALYTICS_LAST_ENTRIES = last_entries
+
+        for func in csv_collectors:
             key = func.__awx_analytics_key__
+            filename = f'{key}.csv'
             try:
-                files = func(last_run, full_path=gather_dir, until=until)
-                if files:
-                    manifest['{}.csv'.format(key)] = func.__awx_analytics_version__
-                if len(files) > num_splits:
-                    num_splits = len(files)
+                slices = [(last_run, until)]
+                if func.__awx_expensive__:
+                    slices = func.__awx_expensive__(key, last_run, until)  # it's ok if this returns a generator
+                for start, end in slices:
+                    files = func(start, full_path=gather_dir, until=end)
+
+                    if not files:
+                        continue
+                    for fpath in files:
+                        tgzfile = package(dest.parent, {filename: (fpath, func.__awx_analytics_version__)}, until)
+                        if tgzfile is not None:
+                            tarfiles.append(tgzfile)
+
+                            if collection_type != 'dry-run':
+                                ship(tgzfile)
```

Comment:
@jbradberry just making sure, we need to send `config.json` with every payload, since it contains the cluster uuid. Can't see it being added here (I can be missing it though :-) )

Reply:
No, you are right.  I'll update to ensure it is included.